api/rest/client/lbclient.go

```diff
@@ -0,0 +1,431 @@
+package client
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	cid "github.com/ipfs/go-cid"
+	shell "github.com/ipfs/go-ipfs-api"
+	files "github.com/ipfs/go-ipfs-files"
+	"github.com/ipfs/ipfs-cluster/api"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+// loadBalancingClient a client to interact with IPFS Cluster APIs.
+// It balances the load by distributing it among peers.
+type loadBalancingClient struct {
+	strategy LBStrategy
+}
+
+// LBStrategy is a strategy to load balance request among clients
+type LBStrategy interface {
+	Next(count int, call func(Client) error) error
```

Comment:
This does not correspond to what I suggested: `Next() Client`

The strategy is not in charge of calling the method, it's just in charge of providing clients in certain order so that the load-balancing client can do that. This way you have ended up with duplicated code blocks in the strategies (creating the clients, handling the errors..).

I guess the `LBStrategy` will also need to implement a `SetClients([]Client)` method as a way to pass in the list of clients as created by the load balancing client.

Reply:
> This way you have ended up with duplicated code blocks in the strategies (creating the clients, handling the errors..).

Changing this to `Next() Client` would also result in similar amount of duplicate code, since to retry the request a `r.retries` you will need to add a loop in every method.

The idea was to have that similar code at a common place.

> The strategy is not in charge of calling the method

If not for this, we would be able to add a `Retry() err` method to the interface, make `Next() Client` and avoid duplicate code from all methods.

Reply:
In the previous version of this code, the load balancer client had a method that did the retry logic for all methods in a single place. That's a problem you had already solved.

Reply:
> I guess the LBStrategy will also need to implement a SetClients([]Client) method as a way to pass in the list of clients as created by the load balancing client.

The clients are added when a strategy is created, I don't see a need for SetClients()

Reply:
You have each strategy creating the clients this duplication client creation code. I don't want users writing alternative strategies to have to create and initialize clients when it can be done for them.

Reply:
https://github.com/ipfs/ipfs-cluster/pull/737/commits/c24f0004fe9ccdf7d365d8709d498c2fdad5a635